// @nekzy/nekzy.js-stream/dist/client/GatewayOpCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayOpCodes = void 0;
var GatewayOpCodes;
(function (GatewayOpCodes) {
    GatewayOpCodes[GatewayOpCodes["dispatch"] = 0] = "dispatch";
    GatewayOpCodes[GatewayOpCodes["heartbeat"] = 1] = "heartbeat";
    GatewayOpCodes[GatewayOpCodes["identify"] = 2] = "identify";
    GatewayOpCodes[GatewayOpCodes["presence_update"] = 3] = "presence_update";
    GatewayOpCodes[GatewayOpCodes["voice_state_update"] = 4] = "voice_state_update";
    GatewayOpCodes[GatewayOpCodes["voice_server_ping"] = 5] = "voice_server_ping";
    GatewayOpCodes[GatewayOpCodes["resume"] = 6] = "resume";
    GatewayOpCodes[GatewayOpCodes["reconnect"] = 7] = "reconnect";
    GatewayOpCodes[GatewayOpCodes["request_guild_members"] = 8] = "request_guild_members";
    GatewayOpCodes[GatewayOpCodes["invalid_session"] = 9] = "invalid_session";
    GatewayOpCodes[GatewayOpCodes["hello"] = 10] = "hello";
    GatewayOpCodes[GatewayOpCodes["heartbeat_ack"] = 11] = "heartbeat_ack";
    GatewayOpCodes[GatewayOpCodes["call_connect"] = 13] = "call_connect";
    GatewayOpCodes[GatewayOpCodes["guild_subscriptions"] = 14] = "guild_subscriptions";
    GatewayOpCodes[GatewayOpCodes["lobby_connect"] = 15] = "lobby_connect";
    GatewayOpCodes[GatewayOpCodes["lobby_disconnect"] = 16] = "lobby_disconnect";
    GatewayOpCodes[GatewayOpCodes["lobby_voice_states_update"] = 17] = "lobby_voice_states_update";
    GatewayOpCodes[GatewayOpCodes["stream_create"] = 18] = "stream_create";
    GatewayOpCodes[GatewayOpCodes["stream_delete"] = 19] = "stream_delete";
    GatewayOpCodes[GatewayOpCodes["stream_watch"] = 20] = "stream_watch";
    GatewayOpCodes[GatewayOpCodes["stream_ping"] = 21] = "stream_ping";
    GatewayOpCodes[GatewayOpCodes["stream_set_paused"] = 22] = "stream_set_paused";
    GatewayOpCodes[GatewayOpCodes["request_guild_application_commands"] = 24] = "request_guild_application_commands";
    GatewayOpCodes[GatewayOpCodes["embedded_activity_launch"] = 25] = "embedded_activity_launch";
    GatewayOpCodes[GatewayOpCodes["embedded_activity_close"] = 26] = "embedded_activity_close";
    GatewayOpCodes[GatewayOpCodes["embedded_activity_update"] = 27] = "embedded_activity_update";
    GatewayOpCodes[GatewayOpCodes["request_forum_unreads"] = 28] = "request_forum_unreads";
    GatewayOpCodes[GatewayOpCodes["remote_command"] = 29] = "remote_command";
})(GatewayOpCodes = exports.GatewayOpCodes || (exports.GatewayOpCodes = {}));
