// @nekzy/nekzy.js-stream/dist/client/voice/VoiceOpCodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceOpCodes = void 0;
var VoiceOpCodes;
(function (VoiceOpCodes) {
    VoiceOpCodes[VoiceOpCodes["identify"] = 0] = "identify";
    VoiceOpCodes[VoiceOpCodes["select_protocol"] = 1] = "select_protocol";
    VoiceOpCodes[VoiceOpCodes["ready"] = 2] = "ready";
    VoiceOpCodes[VoiceOpCodes["heartbeat"] = 3] = "heartbeat";
    VoiceOpCodes[VoiceOpCodes["select_protocol_ack"] = 4] = "select_protocol_ack";
    VoiceOpCodes[VoiceOpCodes["speaking"] = 5] = "speaking";
    VoiceOpCodes[VoiceOpCodes["heartbeat_ack"] = 6] = "heartbeat_ack";
    VoiceOpCodes[VoiceOpCodes["resume"] = 7] = "resume";
    VoiceOpCodes[VoiceOpCodes["hello"] = 8] = "hello";
    VoiceOpCodes[VoiceOpCodes["resumed"] = 9] = "resumed";
    VoiceOpCodes[VoiceOpCodes["sources"] = 12] = "sources";
    VoiceOpCodes[VoiceOpCodes["client_disconnect"] = 13] = "client_disconnect";
    VoiceOpCodes[VoiceOpCodes["session_update"] = 14] = "session_update";
    VoiceOpCodes[VoiceOpCodes["media_sink_wants"] = 15] = "media_sink_wants";
    VoiceOpCodes[VoiceOpCodes["voice_backend_version"] = 16] = "voice_backend_version";
    VoiceOpCodes[VoiceOpCodes["channel_options_update"] = 17] = "channel_options_update";
})(VoiceOpCodes = exports.VoiceOpCodes || (exports.VoiceOpCodes = {}));
